onmt/utils/scoring_utils.py

```diff
@@ -7,78 +7,41 @@
-        elif 'sentencepiece' in opt.transforms:
-            self.type = "sentencepiece"
-        elif 'onmt_tokenize' in opt.transforms:
-            self.type = "pyonmttok"
-            self.tgt_onmttok_kwargs = opt.tgt_onmttok_kwargs
-        else:
-            self.type = None
-        if self.type in ['bpe', 'sentencepiece']:
-            if opt.tgt_subword_model is None:
-                raise ValueError(
-                    "Missing mandatory tokenizer option 'tgt_subword_model'")
-            else:
-                self.model_path = opt.tgt_subword_model
-
-    def build_detokenizer(self):
-        if self.type == "pyonmttok":
-            import pyonmttok
-            self.tgt_detokenizer = pyonmttok.Tokenizer(
-                **self.tgt_onmttok_kwargs)
-        elif self.type == "sentencepiece":
-            import sentencepiece as spm
-            self.tgt_detokenizer = spm.SentencePieceProcessor()
-            self.tgt_detokenizer.Load(self.model_path)
-        elif self.type == "subword-nmt":
-            from subword_nmt.apply_bpe import BPE
-            with open(self.model_path, encoding='utf-8') as tgt_codes:
-                self.tgt_detokenizer = BPE(codes=tgt_codes, vocab=None)
-        else:
-            self.tgt_detokenizer = None
-        return self.tgt_detokenizer
-
-    def _detokenize(self, tokens):
-        if self.type == "pyonmttok":
-            detok = self.tgt_detokenizer.detokenize(tokens)
-        elif self.type == "sentencepiece":
-            detok = self.tgt_detokenizer.DecodePieces(tokens)
-        elif self.type == "subword-nmt":
-            detok = self.tgt_detokenizer.segment_tokens(tokens, dropout=0.0)
-        else:
-            detok = " ".join(tokens)
-        return detok
+from onmt.transforms import get_transforms_cls, make_transforms, TransformPipe
 
 
 class ScoringPreparator():
     """Allow the calculation of metrics via the Trainer's
-     training_eval_handler method"""
+     training_eval_handler method.
+    """
     def __init__(self, vocabs, opt):
         self.vocabs = vocabs
         self.opt = opt
-        self.tgt_detokenizer = Detokenizer(opt)
-        self.tgt_detokenizer.build_detokenizer()
         if self.opt.dump_preds is not None:
             if not os.path.exists(self.opt.dump_preds):
                 os.makedirs(self.opt.dump_preds)
+        self.transforms = opt.transforms
+        transforms_cls = get_transforms_cls(self.transforms)
+        transforms = make_transforms(self.opt, transforms_cls, self.vocabs)
+        self.transform = TransformPipe.build_from(transforms.values())
+
+    def warm_up(self, transforms):
+        self.transforms = transforms
+        transforms_cls = get_transforms_cls(self.transforms)
+        transforms = make_transforms(self.opt, transforms_cls, self.vocabs)
+        self.transform = TransformPipe.build_from(transforms.values())
 
-    def tokenize_batch(self, batch_side, side):
+    def tokenize_batch_side(self, batch, side):
```

Comment:
cf my comment above. maybe we can generalize to 'transform' not only 'tokenization'

Reply:
I don't think "transform" is relevant here, as we do not transform but just retrieve string tokens from ids. But I agree "tokenize" might not be the best name either. Something like "ids_to_tokens" to match the pyonmttok.Vocab wording?

Reply:
what I meant is that a batch of indices is a "transformed" batch. using "tokenize" or "detokenize" here and in scoring utils is a bit misleading, but I'm fine with your suggestion.
As a side note, I am wondering if we could use src_original and tgt_original at some point.